### PR TITLE
skip getaddrinfo thread if host is already resolved, using socket.AI_NUMERIC*

### PR DESCRIPTION
**Pull request by graingert**
should help with #6846

- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/comm/tcp.py [hunk 1]
```diff
@@ -414,11 +414,40 @@ def _check_encryption(self, address, connection_args):
...
+    # If host and port are numeric, then getaddrinfo doesn't block and we can
+    # skip the whole thread thing, which seems worthwhile. So we try first
+    # with the _NUMERIC_ONLY flags set, and then only spawn a thread if that
+    # fails with EAI_NONAME:
+    try:
+        return socket.getaddrinfo(
+            host,
+            port,
+            family,
+            type,
```
Comment (graingert) [inline distributed/comm/tcp.py:427]:
```suggestion
            family=family,
            type=type,
```

File: distributed/comm/tcp.py [hunk 2]
```diff
@@ -414,11 +414,40 @@ def _check_encryption(self, address, connection_args):
             )
 
 
+async def _getaddrinfo(host, port, family, type=socket.SOCK_STREAM):
```
Comment (graingert) [inline distributed/comm/tcp.py:417]:
```suggestion
async def _getaddrinfo(host, port, *, family, type=socket.SOCK_STREAM):
```

File: distributed/comm/tcp.py [hunk 3]
```diff
@@ -414,11 +414,40 @@ def _check_encryption(self, address, connection_args):
             )
 
 
+async def _getaddrinfo(host, port, *, family, type=socket.SOCK_STREAM):
+    # If host and port are numeric, then getaddrinfo doesn't block and we can
+    # skip the whole thread thing, which seems worthwhile. So we try first
```
Comment (fjetter) [inline distributed/comm/tcp.py:419]: From this code alone, it's not apparent what "the whole thread thing" is. I think it's worth mentioning that it is CPython that is dispatching to a thread in this case.

**Review by fjetter — COMMENTED** [review]
If this helps, I'm +1. I think the complexity here is not too bad, even if the patch is not accepted in CPython.

Regarding the OSX slowness. Under which conditions do we experience any slowness? I'm worried about further slowing down our CI since OSX jobs are already the slowest.

Is this "slowness" significant? Would a user working on a mac book notice this when connecting to a cluster?

File: distributed/comm/tcp.py [hunk 4]
```diff
@@ -414,11 +414,40 @@ def _check_encryption(self, address, connection_args):
...
+    # with the _NUMERIC_ONLY flags set, and then only spawn a thread if that
+    # fails with EAI_NONAME:
+    try:
+        return socket.getaddrinfo(
+            host,
+            port,
+            family=family,
+            type=type,
+            socket.AI_NUMERICHOST | socket.AI_NUMERICSERV,
```
Comment (graingert) [inline distributed/comm/tcp.py:428]:
```suggestion
            flags=socket.AI_NUMERICHOST | socket.AI_NUMERICSERV,
```

**Comment by graingert** [comment]
> Is this "slowness" significant? Would a user working on a mac book notice this when connecting to a cluster?

The numeric path should never hit the kernel or a lock, it's supposed to be implemented with just plain bytes parsing
so if it is working as intended, it is executing a couple lines of C code which is basically instantaneous
if it is indeed slow, i.e. a couple of milliseconds would already be very slow, we may need to do something about it


however our choice of what to do when this is slow is limited - we don't know if it's slow until we call it, and we can only know it was too slow after the damage has already been done and we wakeup
we could time the blocking call and throw an error to let the user know raise RuntimeErrro("HEY this is a mythical bug if you can reproduce this go to https://github.com/dask/distributed/issues")

**Comment by graingert** [comment]
```python
def _getaddrinfo_debug(loop, host, port, family, type, proto, flags):
    # based on https://github.com/python/cpython/blob/63140b445e4a303df430b3d60c1cd4ef34f27c03/Lib/asyncio/base_events.py#L833-L856
    t0 = loop.time()
    addrinfo = socket.getaddrinfo(host, port, family, type, proto, flags)
    dt = loop.time() - t0

    if dt >= loop.slow_callback_duration:
        msg = [f"{host}:{port!r}"]
        if family:
            msg.append(f'family={family!r}')
        if type:
            msg.append(f'type={type!r}')
        if proto:
            msg.append(f'proto={proto!r}')
        if flags:
            msg.append(f'flags={flags!r}')
        msg = ', '.join(msg)
        raise RuntimeError(f'Getting address info {msg} took {dt * 1e3:.3f}ms: {addrinfo!r}')
    return addrinfo
```

**Review by gjoseph92 — APPROVED** [review]
> we could time the blocking call and throw an error to let the user know

This is vaguely tempting to me mostly because it costs very little to check, but could be nice to have running in CI. If the error never shows up then it can make us more confident this was the right thing to do.

**Review by fjetter — APPROVED** [review]
Given that this only affects OSX machines, if the problem even exists, I'm OK with this without any additional debugging logs.
The only use case for OSX are LocalClusters and Clients and I don't think slow DNS should be a topic for local clusters.
We do have the "tick warning" implemented and should already get warnings if the loop is blocked for a prolonged period of time on a cluster.
Clients would not be impacted by this since the event loop is most of the time idle anyhow on clients.

**Pull request merged**